### PR TITLE
service: irs: Implement clustering processor

### PR DESCRIPTION
**Pull request by german77**
This infrared image processor is used by `Game Builder Garage` and `Nintendo Labo Toy-Con 01: Variety Kit`. Seems to be fully functional as long there's good contrast with the objects.

image

### CODE REVIEW

**Review by liamwhite — APPROVED** [review]

File: src/core/hle/service/hid/irsensor/clustering_processor.cpp [hunk 1]
```diff
@@ -1,34 +1,265 @@
...
+            if (new_point.y >= height) {
+                continue;
+            }
+            if (GetPixel(data, new_point.x, new_point.y) < current_config.object_intensity_min) {
+                continue;
+            }
+            const ClusteringData cluster = GetPixelProperties(data, new_point.x, new_point.y);
+            current_cluster = MergeCluster(current_cluster, cluster);
+            SetPixel(data, new_point.x, new_point.y, 0);
+            search_points.push({new_point.x, new_point.y});
```
Comment (Morph1984) [inline src/core/hle/service/hid/irsensor/clustering_processor.cpp:142]:
```suggestion
            search_points.emplace(new_point.x, new_point.y);
```

File: src/core/hle/service/hid/irsensor/clustering_processor.cpp [hunk 2]
```diff
@@ -1,34 +1,265 @@
...
+            },
+    };
+}
+
+ClusteringProcessor::ClusteringData ClusteringProcessor::MergeCluster(
+    const ClusteringData a, const ClusteringData b) const {
+    const f32 a_pixel_count = static_cast<f32>(a.pixel_count);
+    const f32 b_pixel_count = static_cast<f32>(b.pixel_count);
+    const f32 pixel_count = a_pixel_count + b_pixel_count;
+    const f32 average_intensitiy =
```
Comment (Morph1984) [inline src/core/hle/service/hid/irsensor/clustering_processor.cpp:175]:
```suggestion
    const f32 average_intensity =
```

File: src/core/hle/service/hid/irsensor/clustering_processor.cpp [hunk 3]
```diff
@@ -1,34 +1,265 @@
...
+    current_config.camera_config.is_negative_used = false;
+    current_config.camera_config.light_target = Core::IrSensor::CameraLightTarget::BrightLeds;
+    current_config.window_of_interest = {
+        .x = 0,
+        .y = 0,
+        .width = width,
+        .height = height,
+    };
+    current_config.pixel_count_min = 3;
+    current_config.pixel_count_max = 0x12C00;
```
Comment (Morph1984) [inline src/core/hle/service/hid/irsensor/clustering_processor.cpp:231]: magic number

File: src/core/hle/service/hid/irsensor/clustering_processor.cpp [hunk 4]
```diff
@@ -1,34 +1,265 @@
...
+        return;
+    }
+
+    next_state = {};
+    const auto camera_data = npad_device->GetCamera();
+    auto filtered_image = camera_data.data;
+
+    RemoveLowIntensityData(filtered_image);
+
+    const std::size_t window_start_x =
```
Comment (lioncash) [inline src/core/hle/service/hid/irsensor/clustering_processor.cpp:56]: You can use auto for these, since the cast already makes the type evident

**Review by liamwhite — APPROVED** [review]

**Pull request merged**